runtime/catalog-karapace/src/main/java/io/aklivity/zilla/runtime/catalog/karapace/internal/KarapaceCatalogHandler.java

```diff
@@ -69,24 +74,41 @@ public KarapaceCatalogHandler(
         this.maxAgeMillis = config.maxAge.toMillis();
         this.event = new KarapaceEventContext(context);
         this.catalogId = catalogId;
+        this.cache = cache;
     }
 
     @Override
     public String resolve(
         int schemaId)
     {
-        String schema;
+        String schema = null;
         if (schemas.containsKey(schemaId))
         {
             schema = schemas.get(schemaId);
         }
         else
         {
-            String response = sendHttpRequest(MessageFormat.format(SCHEMA_PATH, schemaId));
-            schema = response != null ? request.resolveSchemaResponse(response) : null;
-            if (schema != null)
+            CompletableFuture<String> future = cache.get(schemaId);
+            if (future == null)
             {
-                schemas.put(schemaId, schema);
+                future = CompletableFuture.supplyAsync(() ->
+                {
+                    String response = sendHttpRequest(MessageFormat.format(SCHEMA_PATH, schemaId));
+                    return response != null ? request.resolveSchemaResponse(response) : null;
+                });
+            }
+            try
+            {
+                schema = future.get();
+                if (schema != null)
+                {
+                    cache.put(schemaId, future);
+                    schemas.put(schemaId, schema);
+                }
+            }
+            catch (ExecutionException | InterruptedException e)
+            {
+                // TODO: log an event
```

Comment:
There is still a race here, between `cache.get(schemaId)` and `cache.put(schemaId, future)`.

To solve the race, we need to use one of the methods on `ConcurrentMap` that is not available on `Map`, such as:
```
CompletableFuture<> future = cache.get(schemaId);
if (future == null)
{
    CompletableFuture<> newFuture = new CompletableFuture<String>();
    future = cache.putIfAbsent(schemaId, newFuture);
    if (future == null)
    {
        // send request and complete newFuture
        future = newFuture;
    }
}
assert future != null;
schema = future.get();
if (schema != null)
{
    schemas.put(schemaId, schema);
}
```

Reply:
Note, we also need to prevent unnecessary concurrent requests to resolve the schemaIds, such as `latest`, same as for resolving the schemas themselves from schemaIds.